src/modm/platform/dma/samx7x/base_impl.hpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2023, Christopher Durand
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+namespace modm::platform
+{
+
+constexpr void
+DmaBase::ViewBase::setDataLength(std::size_t length)
+{
+	uint32_t& mbr_ubc = data_[1];
+	mbr_ubc = (mbr_ubc & ~0xFF'FF'FFu) | (length & 0xFF'FF'FFu);
+}
+
+constexpr std::size_t
+DmaBase::ViewBase::dataLength() const
+{
+	const uint32_t& mbr_ubc = data_[1];
+	return mbr_ubc & 0xFF'FF'FFu;
+}
+
+constexpr void
+DmaBase::View0Dst::setDestinationAddress(uintptr_t address)
+{
+	data_[2] = address;
+}
+
+constexpr void
+DmaBase::View0Dst::setDestinationAddress(volatile void* data)
+{
+	data_[2] = reinterpret_cast<uint32_t>(data);
+}
+
+constexpr uintptr_t
+DmaBase::View0Dst::destinationAddress() const
+{
+	static_assert(sizeof(uint32_t) == sizeof(uintptr_t));
+	return data_[2];
+}
+
+constexpr void
+DmaBase::View0Src::setSourceAddress(uintptr_t address)
+{
+	data_[2] = address;
+}
+
+constexpr void
+DmaBase::View0Src::setSourceAddress(const volatile void* data)
+{
+	data_[2] = reinterpret_cast<uint32_t>(data);
+}
+
+constexpr uintptr_t
+DmaBase::View0Src::sourceAddress() const
+{
+	static_assert(sizeof(uint32_t) == sizeof(uintptr_t));
+	return data_[2];
+}
+
+constexpr void
+DmaBase::View1::setDestinationAddress(uintptr_t address)
+{
+	data_[3] = address;
+}
+
+constexpr void
+DmaBase::View1::setDestinationAddress(volatile void* data)
+{
+	data_[3] = reinterpret_cast<uint32_t>(data);
+}
+
+constexpr uintptr_t
+DmaBase::View1::destinationAddress() const
+{
+	static_assert(sizeof(uint32_t) == sizeof(uintptr_t));
+	return data_[3];
+}
+
+constexpr void
+DmaBase::View2::setConfiguration(ChannelConfig_t configuration)
+{
+	data_[4] = configuration.value;
+}
+
+constexpr auto
+DmaBase::View2::configuration() const -> ChannelConfig_t
```

Comment:
```suggestion
constexpr ChannelConfig_t
DmaBase::View2::configuration() const
```
Consistency?

Reply:
Nope, you suggestion does not compile. Would be:

```
constexpr DmaBase::View2::ChannelConfig_t
DmaBase::View2::configuration() const
```
or something like that and that's less readable in my opinion. To avoid that I have used the trailing return type which is looked up in the class scope.